Add ReverseAddrIP function

Felixoid opened the pull request:
Hello Meik, thank you for a great project.

To avoid unnecessary conversions from `dns.A -> string -> dns.ReverseAddr(string)` I've implemented ReverseAddrIP function.

Here you can see the benchmark results:

```
go test -bench BenchmarkReverse -benchmem . -benchtime 10000000x
goos: linux
goarch: amd64
pkg: github.com/miekg/dns
cpu: AMD Ryzen 7 4800H with Radeon Graphics         
BenchmarkReverseAddr/IP4-16      	10000000	      180.3 ns/op	     40 B/op	      2 allocs/op
BenchmarkReverseAddr/IP6-16      	10000000	      234.4 ns/op	     96 B/op	      2 allocs/op
BenchmarkReverseAddrIP/IP4-16    	10000000	       96.96 ns/op	     24 B/op	      1 allocs/op
BenchmarkReverseAddrIP/IP6-16    	10000000	      142.1 ns/op	     80 B/op	      1 allocs/op
BenchmarkReverseAddrFromIP/IP4-16         	10000000	      222.3 ns/op	     56 B/op	      3 allocs/op
BenchmarkReverseAddrFromIP/IP6-16         	10000000	      342.1 ns/op	    112 B/op	      3 allocs/op
PASS
ok  	github.com/miekg/dns	15.107s
```

tmthrgd commented on defaults.go:
```diff
@@ -333,6 +333,12 @@ func ReverseAddr(addr string) (arpa string, err error) {
+
+// ReverseIP returns the in-addr.arpa. or ip6.arpa. hostname of the IP
+// address suitable for reverse DNS (PTR) record lookups.
+func ReverseIP(ip net.IP) (arpa string) {
```
`(arpa string)` -> `string`

tmthrgd commented on defaults.go:
```diff
@@ -333,6 +333,12 @@ func ReverseAddr(addr string) (arpa string, err error) {
+}
+
+// ReverseIP returns the in-addr.arpa. or ip6.arpa. hostname of the IP
+// address suitable for reverse DNS (PTR) record lookups.
```
This should document that `ip` must be either an IPv4 or IPv6 address.

tmthrgd replied:
I’d probably also like it to return an empty string if it’s invalid. I think `ip.To6() == nil` should be the right check.

Felixoid replied:
When you say so regarding check for `ip.To16`, I feel like the function should return an error as well

tmthrgd commented:
I’m fine with this, it makes sense to me. Though I think `ReverseAddrIP` might be a better name.

Felixoid commented:
Thank you for review, I've added the changes

Felixoid commented:
@tmthrgd and @miekg, anything preventing it from merging?

Felixoid commented:
Well, I hope you at least have a lot of fun on vacations :+1: :heart:

Felixoid commented:
@miekg and @tmthrgd, any hopes to get it reviewed?

miekg commented:
There is also ReverseAddr ? So now ewe have two? The benchmarks show ns/op but you claim memory efficiency, also don't really think that's important...
The commit message are rather sparse and don't have any of the comments you put in the PR text

Felixoid commented:
Hey Miek.

> There is also ReverseAddr ? So now we have two?

Yes. Somebody may find it useful to have ReverseAddr still. So I don't feel comfortable breaking the backward compartibility.

> The benchmarks show ns/op but you claim memory efficiency, also don't really think that's important...

```
go test -bench BenchmarkReverse -benchmem . -benchtime 10000000x
goos: linux
goarch: amd64
pkg: github.com/miekg/dns
cpu: AMD Ryzen 7 4800H with Radeon Graphics         
BenchmarkReverseAddr/IP4-16      	10000000	      180.3 ns/op	     40 B/op	      2 allocs/op
BenchmarkReverseAddr/IP6-16      	10000000	      234.4 ns/op	     96 B/op	      2 allocs/op
BenchmarkReverseAddrIP/IP4-16    	10000000	       96.96 ns/op	     24 B/op	      1 allocs/op
BenchmarkReverseAddrIP/IP6-16    	10000000	      142.1 ns/op	     80 B/op	      1 allocs/op
BenchmarkReverseAddrFromIP/IP4-16         	10000000	      222.3 ns/op	     56 B/op	      3 allocs/op
BenchmarkReverseAddrFromIP/IP6-16         	10000000	      342.1 ns/op	    112 B/op	      3 allocs/op
PASS
ok  	github.com/miekg/dns	15.107s
```

I'd rather agree with you. But the new function looks useful to me since it avoids the unnecessary converting from `net.IP` address to string and back to `net.IP`. To show what I mean I've added BenchmarkReverseAddrFromIP

> The commit message are rather sparse and don't have any of the comments you put in the PR text

Updated the commit and rebased it on top of the current master :+1:

Felixoid commented:
Ok, I didn't get the answer, but from the gut feeling, it looks like the current approach is better than adding an extra function.

tmthrgd commented on defaults.go:
```diff
@@ -326,13 +327,23 @@ func CanonicalName(s string) string {
+// address suitable for reverse DNS (PTR) record lookups. It accepts string or net.IP
+// as addr parameter. Error is returned if it's unable to parse IP address from string,
+// neither IPv4 nor IPv6 is used, or other but string of net.IP object is used
+func ReverseAddr(addr interface{}) (arpa string, err error) {
```
This isn't backwards compatible so we can't make this change.

tmthrgd reviewed (changes requested):
This isn't backwards compatible.
